src/i_3dsound.c

```diff
@@ -0,0 +1,312 @@
+{
+    ALfloat position[3];
+    ALfloat velocity[3];
+    ALfloat direction[3];
+    boolean point_source;
+} oal_source_params_t;
+
+static boolean I_3D_InitSound(void)
+{
+    return I_OAL_InitSound(true);
+}
+
+static boolean I_3D_ReinitSound(void)
+{
+    return I_OAL_ReinitSound(true);
+}
+
+static void CalcPitchAngle(const player_t *player, angle_t *pitch)
+{
+    fixed_t lookdir, slope;
+
+    // Same as R_SetupFrame() in r_main.c.
+    lookdir = player->lookdir / MLOOKUNIT + player->recoilpitch;
+
+    if (lookdir == 0)
+    {
+        *pitch = 0;
+        return;
+    }
+
+    if (lookdir > LOOKDIRMAX)
+        lookdir = LOOKDIRMAX;
+    else if (lookdir < -LOOKDIRMAX)
+        lookdir = -LOOKDIRMAX;
+
+    slope = (lookdir << FRACBITS) / VIEWDIST;
+    if (slope < 0)
+    {
+        *pitch = (ANGLE_MAX - tantoangle[-slope >> DBITS]) >> ANGLETOFINESHIFT;
+    }
+    else
+    {
+        *pitch = tantoangle[slope >> DBITS] >> ANGLETOFINESHIFT;
+    }
+}
+
+static void CalcListenerParams(const mobj_t *listener, oal_listener_params_t *lis)
+{
+    const player_t *player = listener->player;
+    const angle_t yaw = listener->angle >> ANGLETOFINESHIFT;
+    angle_t pitch;
+
+    lis->position[0] = FIXED_TO_ALFLOAT(listener->x);
+    lis->position[1] = FIXED_TO_ALFLOAT(player->viewz);
+    lis->position[2] = FIXED_TO_ALFLOAT(-listener->y);
+
+    lis->velocity[0] = FIXED_TO_ALFLOAT(listener->momx) * TICRATE;
+    lis->velocity[1] = FIXED_TO_ALFLOAT(listener->momz) * TICRATE;
+    lis->velocity[2] = FIXED_TO_ALFLOAT(-listener->momy) * TICRATE;
+
+    CalcPitchAngle(player, &pitch);
+    if (pitch == 0)
+    {
+        // "At" vector after yaw rotation.
+        lis->orientation[0] = FIXED_TO_ALFLOAT(finecosine[yaw]);
+        lis->orientation[1] = 0.0f;
+        lis->orientation[2] = FIXED_TO_ALFLOAT(-finesine[yaw]);
+
+        // "Up" vector doesn't change.
+        lis->orientation[3] = 0.0f;
+        lis->orientation[4] = 1.0f;
+        lis->orientation[5] = 0.0f;
+    }
+    else
+    {
+        // "At" vector after yaw and pitch rotations.
+        lis->orientation[0] = FIXED_TO_ALFLOAT(finecosine[yaw] * finesine[pitch]);
+        lis->orientation[1] = FIXED_TO_ALFLOAT(finecosine[pitch]);
+        lis->orientation[2] = FIXED_TO_ALFLOAT(-finesine[yaw] * finesine[pitch]);
+
+        // "Up" vector after yaw and pitch rotations.
+        lis->orientation[3] = FIXED_TO_ALFLOAT(-finecosine[yaw] * finecosine[pitch]);
+        lis->orientation[4] = FIXED_TO_ALFLOAT(finesine[pitch]);
+        lis->orientation[5] = FIXED_TO_ALFLOAT(finesine[yaw] * finecosine[pitch]);
+    }
+}
+
+static void CalcSourceParams(const mobj_t *listener, const mobj_t *source,
+                             oal_source_params_t *src)
+{
+    src->position[0] = FIXED_TO_ALFLOAT(source->x);
+    src->position[2] = FIXED_TO_ALFLOAT(-source->y);
+
+    // Treat monsters and projectiles as point sources.
+    src->point_source = source->info && source->info->actualheight;
```

Comment:
Could you please use parentheses when using boolean expressions in assignments or `return()`s?

Reply:
Fixed by https://github.com/fabiangreffrath/woof/pull/1118/commits/1ed469b6beb0d737ffb298475166aad1b28ee9b3